File: src/medium-137/solution.go
package medium137

func singleNumber(nums []int) int {
	seenOnce, seenTwice := 0, 0

	for _, num := range nums {
		// golang 中使用前缀 ^ 表示按位非，而不是前缀 ~
		seenOnce = ^seenTwice & (seenOnce ^ num)
		seenTwice = ^seenOnce & (seenTwice ^ num)
	}

	return seenOnce
}

File: src/medium-137/solution.ts
/**
 * @ref https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#bitwise_operators
 */
export function singleNumber(nums: number[]): number {
  let seenOnce = 0,
    seenTwice = 0

  nums.forEach((num) => {
    seenOnce = ~seenTwice & (seenOnce ^ num)
    seenTwice = ~seenOnce & (seenTwice ^ num)
  })

  return seenOnce
}
